v2/pkg/subscraping/types.go

```diff
@@ -20,54 +20,35 @@ type Source interface {
 	// which contains the extractor for subdomains, http client
 	// and other stuff.
 	Run(context.Context, string, *Session) <-chan Result
-	// Name returns the name of the source
+	// Name returns the name of the source. It is preferred to use lower case names.
 	Name() string
+
+	// IsDefault returns true if the current source should be
+	// used as part of the default execution.
+	IsDefault() bool
+
+	// HasRecursiveSupport returns true if the current source
+	// accepts subdomains (e.g. subdomain.domain.tld),
+	// not just root domains.
+	HasRecursiveSupport() bool
+
+	// NeedsKey returns true if the source requires an API key
+	NeedsKey() bool
+
+	AddApiKeys([]string)
```

Comment:
couldn't we have the API keys stored somewhere instead of having a method that stores these values per provider? it just seems more elegant to have them stored somewhere as we had previously and picked randomly without this being done at time of initialization

Reply:
No, that's the whole point. Sources should be self-contained. About elegance, I believe my approach is far more elegant than the previous one, not to mention that it's significantly less error prone and easier to add new sources.

Reply:
i think there a few major problems with this approach. I like the AddApiKeys function but the way we implement the addition is an issue, mainly because the api keys are stored as global variables. which means if we want to use multiple  instances of source we cannot do it. Instead it should be stored as a variable in source itself and probably that'll be better. Also i see the function code related to API keys has been copy pasted a lot so maybe that is something we should fix as well

Reply:
The keys are now moved inside the `Source` struct as discussed. The code related to API keys is not copy-pasted, but uses a generic method that enables naming the key values individually for each source.